### ISSUE TITLE
[Beetle PSX HW] Lightgun API is broken

### ISSUE THREAD

**Issue by Phiben** [libretro/beetle-psx-libretro#501]
Since the release of the Lightgun API  #271 by @hiddenasbestos  back in 2017 it never worked in Beetle PSX HW.

It kind of works in the non-HW version of the core but only at native resolution. If the resolution is higher than native, the gun alignment is completely off.

In Beetle PSX HW (OpenGL and Vulkan) it simply doesn't work at all. Neither at native resolution nor at upscaled resolutions.

**Comment by chasburger** [comment]
Yeah I've run into this as well. Lightgun functionality in Bettle PSX HW when using the OpenGL or Vulkan driver is non-existent. The trigger and button inputs are sent, but the cursor is not drawn..

I can confirm that using higher than native internal resolution in the non-HW version messes with the alignment as well.

I will say that the lightgun API functions in Beetle PSX HW if you use DX11 or 12 though. Cursor is drawn and everything works as it should. So I just use that as a workaround for now. Slightly irritating to have to load another video driver and restart the RetroArch frontend, but it's not the end of the world.

**Comment by rz5** [comment]
@chasburger Beetle PSX HW only has hardware renderers for OpenGL and Vulkan. If you're using RetroArch with a video driver different from gl, glcore and vulkan, that's solely software rendering.

**Comment by ggdrt** [comment]
Diagnosis of the alignment issues: https://github.com/libretro/beetle-psx-libretro/issues/471#issuecomment-558778374

As for the lightgun not showing cursor for hardware renderers, that's probably going to require an RSX interface hook in `mednafen/psx/frontio.cpp` for either the `crosshair_plot()` or `DrawCrosshairs()` function. These draw functions also need to be adjusted to scale the cursor's size up proportionally for the software renderer at higher internal resolutions.

**Comment by Darknior** [comment]
Hi everyone, some news about this problem ? Some solution to solve it ?
Thanks

**Comment by Tovarichtch** [comment]
I've tested in software rendering and indeed, it works. Hardware rendering is the problem here, whatever the API used. It would be great to fix this, since it's the best core with all the light guns required to play all gun games (including Area 51 which is not possible to play without the Justifier, and it has no GunCon conversion patch).

**Comment by SeongGino** [comment]
Can confirm that this still seems to be broken with hardware rendering specifically, as it works totally fine in software mode (with all the performance caveats that it brings).

At least the few Justifier-only games on PS1 are predominantly 2D in nature (so Lethal Enforcers for example wouldn't benefit much, if at all), though this might suck a little for the likes of Crypt Killer, Project Horned Owl and Die Hard Trilogy which could look much better.

For now, the workaround will have to be to run Beetle-HW for those Justifier-only games in software mode, and run GunCon games in SwanStation if you desire HW rendering.
